### PR TITLE
Support unix_timestamp and to_unix_timestamp with non-UTC timezones (non-DST)

### PR DESCRIPTION
**Pull request by winningsix**
This fixes https://github.com/NVIDIA/spark-rapids/issues/9815,  https://github.com/NVIDIA/spark-rapids/issues/9606 and https://github.com/NVIDIA/spark-rapids/issues/10006

TODOs:
- [x] Support ```ImprovedTimestamp``` path
- [x] Fix mis-matched CPU/GPU execution result

### CODE REVIEW

File: integration_tests/src/main/python/date_time_test.py [hunk 1]
```diff
@@ -262,6 +262,13 @@ def test_unsupported_fallback_to_unix_timestamp(data_gen):
         "to_unix_timestamp(a, b)"),
         "ToUnixTimestamp")
 
+# TODO: has another test for this called test_unix_timestamp
+@pytest.mark.parametrize('data_gen', date_n_time_gens, ids=idfn)
```
Comment (res-life) [inline integration_tests/src/main/python/date_time_test.py:266]: We should also test string gen.
Please refer to the Product plan doc, it requires String data type.

**Comment by winningsix** [comment]
build

**Comment by sperlingxx** [comment]
LGTM

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 2]
```diff
@@ -824,25 +824,42 @@ abstract class GpuToTimestamp
...
+        } else {
+          parseStringAsTimestamp(
+            lhs,
+            sparkFormat,
+            strfFormat,
+            DType.TIMESTAMP_MICROSECONDS,
+            failOnError)
+        }
+      case _: DateType =>
+        timeZoneId match {
```
Comment (res-life) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:846]: Use `zoneId` defined in `TimeZoneAwareExpression`
```
@transient lazy val zoneId: ZoneId = DateTimeUtils.getZoneId(timeZoneId.get)
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 3]
```diff
@@ -892,8 +909,20 @@ abstract class GpuToTimestampImproved extends GpuToTimestamp {
           failOnError)
       }
     } else if (lhs.dataType() == DateType){
-      lhs.getBase.asTimestampSeconds()
-    } else { // Timestamp
+      timeZoneId match {
```
Comment (res-life) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:912]: Use `zoneId` defined in `TimeZoneAwareExpression`
```
@transient lazy val zoneId: ZoneId = DateTimeUtils.getZoneId(timeZoneId.get)
```

File: integration_tests/src/main/python/date_time_test.py [hunk 4]
```diff
@@ -245,10 +245,10 @@ def test_dayofyear(data_gen):
             lambda spark : unary_op_df(spark, data_gen).select(f.dayofyear(f.col('a'))))
 
 @pytest.mark.parametrize('data_gen', date_n_time_gens, ids=idfn)
-@allow_non_gpu(*non_utc_allow)
 def test_unix_timestamp(data_gen):
     assert_gpu_and_cpu_are_equal_collect(
-            lambda spark : unary_op_df(spark, data_gen).select(f.unix_timestamp(f.col('a'))))
```
Comment (res-life) [inline integration_tests/src/main/python/date_time_test.py:251]: Add cases for not supported Time Zone.

e.g.:
```
@pytest.mark.skipif(not is_supported_time_zone(), reason="not all time zones are supported now, refer to https://github.com/NVIDIA/spark-rapids/issues/6839, please update after all time zones are supported")
def test_from_unixtime(data_gen, date_format):

@pytest.mark.skipif(is_supported_time_zone(), reason="not all time zones are supported now, refer to https://github.com/NVIDIA/spark-rapids/issues/6839, please update after all time zones are supported")
def test_from_unixtime_fall_back(data_gen, date_format):
```

And test TZs locally:
export TZ=Iran
export TZ= America/Los_Angeles

Comment (winningsix) [reply integration_tests/src/main/python/date_time_test.py:251]: updated

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala (hunk 2 above)
Comment (winningsix) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:846]: Still ```timeZoneId``` to match non case but use it later.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala (hunk 3 above)
Comment (winningsix) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:911]: ditto

**Comment by res-life** [comment]
LGTM

**Review by res-life — DISMISSED** [review]

**Comment by winningsix** [comment]
build

**Review by NVnavkumar — APPROVED** [review]
LGTM. Eventually, need to do cleanup based on https://github.com/NVIDIA/spark-rapids-jni/issues/1643

File: integration_tests/src/main/python/date_time_test.py [hunk 5]
```diff
@@ -398,12 +404,13 @@ def invalid_date_string_df(spark):
     return spark.createDataFrame([['invalid_date_string']], "a string")
 
 @pytest.mark.parametrize('ansi_enabled', [True, False], ids=['ANSI_ON', 'ANSI_OFF'])
+@pytest.mark.parametrize('non_utc_timezone_enabled', [True, False], ids=['NOT_UTC_ON', 'NOT_UTC_OFF'])
 @pytest.mark.parametrize('data_gen,date_form', str_date_and_format_gen, ids=idfn)
-@allow_non_gpu(*non_utc_allow)
-def test_string_to_unix_timestamp(data_gen, date_form, ansi_enabled):
+@allow_non_gpu('ProjectExec')
+def test_string_to_unix_timestamp(data_gen, date_form, ansi_enabled, non_utc_timezone_enabled):
```
Comment (revans2) [inline integration_tests/src/main/python/date_time_test.py:410]: This parameter is not used and nonUTC.enabled is going away.

File: integration_tests/src/main/python/date_time_test.py [hunk 6]
```diff
@@ -398,12 +404,13 @@ def invalid_date_string_df(spark):
     return spark.createDataFrame([['invalid_date_string']], "a string")
 
 @pytest.mark.parametrize('ansi_enabled', [True, False], ids=['ANSI_ON', 'ANSI_OFF'])
+@pytest.mark.parametrize('non_utc_timezone_enabled', [True, False], ids=['NOT_UTC_ON', 'NOT_UTC_OFF'])
 @pytest.mark.parametrize('data_gen,date_form', str_date_and_format_gen, ids=idfn)
-@allow_non_gpu(*non_utc_allow)
-def test_string_to_unix_timestamp(data_gen, date_form, ansi_enabled):
+@allow_non_gpu('ProjectExec')
```
Comment (revans2) [inline integration_tests/src/main/python/date_time_test.py:409]: I don't want to have a hard coded ProjectExec that does not know if we should fall back to the CPU or not!.

File: integration_tests/src/main/python/date_time_test.py [hunk 7]
```diff
@@ -412,12 +419,13 @@ def test_string_to_unix_timestamp_ansi_exception():
         conf=ansi_enabled_conf)
 
 @pytest.mark.parametrize('ansi_enabled', [True, False], ids=['ANSI_ON', 'ANSI_OFF'])
+@pytest.mark.parametrize('non_utc_timezone_enabled', [True, False], ids=['NOT_UTC_ON', 'NOT_UTC_OFF'])
 @pytest.mark.parametrize('data_gen,date_form', str_date_and_format_gen, ids=idfn)
-@allow_non_gpu(*non_utc_allow)
-def test_string_unix_timestamp(data_gen, date_form, ansi_enabled):
+@allow_non_gpu('ProjectExec')
```
Comment (revans2) [inline integration_tests/src/main/python/date_time_test.py:424]: Here too.

**Review by revans2 — CHANGES_REQUESTED** [review]
I know this just got merged, but we need to go back and fix the tests.

**Pull request merged**